### PR TITLE
threading branch

### LINKED ISSUE #10772: Help sought on updating `threading` branch -- LLVM issues
I've been trying to update the `threading` branch to get it on par with `master`, and running into a problem that I need help with.  The symptom is during a build, I get:
```
    LINK usr/bin/julia
    JULIA usr/lib/julia/sys0.o
LLVM ERROR: Cannot select: 0x212a650: i64 = X86ISD::WrapperRIP 0x212a548 [ORD=6] [ID=76] dbg:boot.jl:277
  0x212a548: i64 = TargetGlobalTLSAddress<%jl_value_t*** @jl_pgcstack> 0 [TF=10] [ORD=6] [ID=45] dbg:boot.jl:277
In function: julia__new_0
make[1]: *** [/localdisk/adrobiso/tmp/julia/usr/lib/julia/sys0.o] Error 1
make: *** [julia-sysimg-release] Error 2
```
@keno mentioned a while back that the problem might be that the right options are not being passed to LLVM.  But the options appear to be the same as the working `threading` branch, as far as I can tell from poking around with gdb.  Alas the call chain goes through a pattern-matching interpreter that I don't understand.

I've put a copy of my current work in branch `adr/theading`.  It's a merge of two branches:
* `threading`
* `adr/snapshot`, which is a snapshot of `master` that I took yesterday.
The number of threads defaults to `1`, so the problem is likely not a race condition.  (indeed, I still have work to do on `gc.c` to make it thread-safe.)  Any suggestions on how to fix or debug the LLVM error would be appreciated.

### PR DESCRIPTION
**Pull request by vtjnash**
this closes https://github.com/JuliaLang/julia/issues/10772 and superceeds the other threading branches

the goal of this is to have the threading support off by default (controlled by a flag in julia.h), but to have the code building and passing tests either way; and to get this merged into master sooner rather than later.

### CODE REVIEW

**Comment by timholy** [comment]
Wow, really exciting stuff!

**Comment by johnmyleswhite** [comment]
It's like Christmas day over here

**Comment by tkelman** [comment]
Everything with an Intel license needs to be spelled out in `LICENSE.md` and `contrib/add_license_to_files.jl`.

File: Make.inc [hunk 1]
```diff
@@ -496,6 +494,8 @@ else
 ifneq ($(XC_HOST),)
 XC_HOST := $(ARCH)$(shell echo $(XC_HOST) | sed "s/[^-]*\(.*\)$$/\1/")
 MARCH := $(subst _,-,$(ARCH))
+else # insert ARCH into HOST
+XC_HOST := $(ARCH)$(shell echo $(BUILD_MACHINE) | sed "s/[^-]*\(.*\)$$/\1/")
```
Comment (tkelman) [inline Make.inc]: what's the purpose of always setting `XC_HOST` even when not cross-compiling?

File: test/perf/threads/lbm3d/lbm3d.jl [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                     	  [-1  0  1], [-1  0 -1], [ 0  1  1],
+                     	  [ 0  1 -1], [ 0 -1  1], [ 0 -1 -1]);
+
+const fourths = tuple([ 6, 8, 9,12,13], [ 7,10,11,14,15],
+                      [ 4, 8,10,16,17], [ 5, 9,11,18,19],
+                      [ 2,12,14,16,18], [ 3,13,15,17,19])
+
+
+function relax!(F, UX, UY, UZ, nx, ny, nz, deltaU, t1D, t2D, t3D, sSQU)
+  tid = threadid()
```
Comment (tkelman) [inline test/perf/threads/lbm3d/lbm3d.jl]: 4-space this file

File: Make.inc (hunk 1 above)
Comment (vtjnash) [inline Make.inc]: if you explicitly set `ARCH`, then it assumes you meant to try to do a cross-compile. without this, libunwind was getting confused by the mismatch between `--host` and `CC=gcc -m32`.

Comment (tkelman) [inline Make.inc]: Good to know, thanks. Wouldn't that have happened independently of the threading branch though?

Comment (vtjnash) [inline Make.inc]: sure. i don't go around randomly testing build configurations however, so i only notice these issues when building branches.

**Comment by Keno** [comment]
If they're gonna be part of the base library, they should have our standard license header.

**Comment by Keno** [comment]
@kpamnany Might be good to just run that by Intel Legal. Shouldn't be a problem. Most other open source projects have the same policy (e.g. LLVM which Intel contributes to heavily).

**Comment by kpamnany** [comment]
Have inquired and will get back on changing the license to MIT.

**Comment by kpamnany** [comment]
Confusing build errors coming from code under LLVM38. Calling ``createTypeBasedAAWrapperPass`` but it doesn't exist in ``deps/llvm-svn``. How was this compiling previously?

**Comment by StefanKarpinski** [comment]
llvm-svn is a moving target?

**Comment by kpamnany** [comment]
This branch of ``llvm-svn`` shouldn't be moving (elegantly named ``tlsrebaseagainagain``).

**Comment by Keno** [comment]
This PR is rebased against master. Does it still include that?

**Comment by kpamnany** [comment]
Yes.
```
~/13410/deps/srccache/llvm-svn$ git branch
* jn/tlsrebaseagainagain
```

**Comment by Keno** [comment]
Ok, then it's probably the other way around. Master is head in llvm version of threading, so we might need to rebase again.

**Comment by vtjnash** [comment]
In order to make this merge-ready with master, I removed the Make.user customizations, including the llvm branch specification.

**Comment by kpamnany** [comment]
Right. It builds fine with ``JL_ENABLE_THREADING`` turned off, and with whatever is LLVM default.

But when you turn on ``JL_ENABLE_THREADING`` and add in llvm-svn into Make.user as we need it for threading, the build fails with multiple errors in ``codegen.cpp``, all under ``#ifdef LLVM38``.

**Comment by Keno** [comment]
I think you might just have a stale branch in deps/srccache, which is causing your build failures.

**Comment by kpamnany** [comment]
This is a fresh clone, to test exactly this.

**Comment by vtjnash** [comment]
the primary blocker for me on this PR right now is that gcc4/5 doesn't compile Julia correctly with `-m32 -march=pentium4`, so the SSE2 instruction `pause` cannot be used. In particular, in the expression in `jl_compute_field_offsets`:
```c
            size_t alsz = (((sz) + (al-1)) & (-al));
            if (alsz > sz)
                st->haspadding = 1;
```
in the assembly output, it has effectively turned that conditional into `if (sz > sz)`

I don't think I see any undefined behavior here according to the C standard, so I filed a bugreport against gcc: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67814

**Comment by tkelman** [comment]
Any hope of a workaround? GCC bug is not a good time scale on this.

**Comment by Keno** [comment]
inline assembly?

**Comment by kpamnany** [comment]
I still see those compile errors in ``codegen.cpp``. What's wrong there?

**Comment by vtjnash** [comment]
@kpamnany can you change the license on the files to the standard text ("# This file is a part of Julia. License is MIT: http://julialang.org/license\n\n")? then I think this should be ready to merge.

**Comment by kpamnany** [comment]
Done. But I still see those compile errors with ``JULIA_ENABLE_THREADING`` turned on.

**Comment by Keno** [comment]
Just needs a rebase of the threading patches on top of llvm svn.

**Comment by Keno** [comment]
I'd the locking implementation to be available even if threads a are disabled to separate the profiler from trying to unwind while we're modifying libunwind's internal data structures.

**Comment by Keno** [comment]
But that can come after merging.

File: base/locks.jl [hunk 3]
```diff
@@ -0,0 +1,138 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+export SpinLock, Mutex, init_lock!, destroy_lock!, lock!, trylock!, unlock!
+
+abstract Lock
```
Comment (tkelman) [inline base/locks.jl]: should this be called `AbstractLock` ?

**Comment by kpamnany** [comment]
Expect lots of input and suggestions on desired interface and functionality. These can go on top of the list for after merge. :)

**Comment by tkelman** [comment]
What are the licenses on the original code for `stockcorr_vec.m` and `lbm3d.m` ? I guess we already had the stockcorr benchmarks at https://github.com/JuliaLang/julia/blob/master/test/perf/kernel/stockcorr.m, but did anyone ever ask the original author for permission?

File: base/cartesian.jl [hunk 4]
```diff
@@ -98,6 +98,19 @@ function _nall(N::Int, criterion::Expr)
     Expr(:&&, conds...)
 end
 
+# Check whether any of variables i1, i2, ... satisfy criterion
+macro nany(N, criterion)
+    _nany(N, criterion)
+end
+
+function _nany(N::Int, criterion::Expr)
```
Comment (IainNZ) [inline base/cartesian.jl]: Tests?

File: base/locks.jl [hunk 5]
```diff
@@ -0,0 +1,138 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
```
Comment (amitmurthy) [inline base/locks.jl]: There is already a `base/lock.jl`. Shouldn't the two be merged?

**Comment by tkelman** [comment]
Would need the following patch to build under msvc: https://gist.github.com/676fffb916a7d67f4a59

Anyone opposed to committing that here?

File: src/options.h [hunk 6]
```diff
@@ -17,11 +17,8 @@
 
 // codegen options ------------------------------------------------------------
 
-// (Experimental) codegen support for thread-local storage
-// #define CODEGEN_TLS
-
 // (Experimental) Use MCJIT ELF, even where it's not the native format
-// #define FORCE_ELF
+#define FORCE_ELF
```
Comment (vtjnash) [inline src/options.h]: this line might need to be reverted before merging. @Keno

File: base/locks.jl (hunk 5 above)
Comment (kpamnany) [inline base/locks.jl]: Yes. I'm working on the interaction between threads and tasks, which is also needed for I/O. When that is properly designed, these will get merged.

**Comment by kpamnany** [comment]
@tkelman: I emailed the author and he has updated the post at http://www.walkingrandomly.com/?p=3604 to state that the code is licensed under the 3 clause BSD license. He's also asked for his name and a link as credit; I've added those.

File: base/cartesian.jl (hunk 4 above)
Comment (kpamnany) [inline base/cartesian.jl]: I'd add a couple, but I can't find tests for any of the existing macros in ``Cartesian``.

**Comment by tkelman** [comment]
@kpamnany great, that is good news, thanks for following up on that (and to the author for responding quickly). I think we do need to put the license text in accompanying the source code, either inline or in the same directory.

**Comment by ViralBShah** [comment]
@kpamnany Thanks. I think the code had come up on the mailing list, and I had put it in our benchmarks. @tkelman Thanks for noticing.

**Comment by vtjnash** [comment]
please use the llvm-svn branch `jn/tls37` now. i'm getting ready to merge this, so let me know if there remaining problems.

**Comment by tkelman** [comment]
`lbm3d.m` is still missing a license

**Comment by kpamnany** [comment]
@vtjnash: with threading turned on, there are no compile errors with ``jn/tls37``, and threads seem to work fine.

@tkelman: I've emailed the author of LBM. Awaiting a response.

File: src/support/dtypes.h [hunk 7]
```diff
@@ -137,6 +137,14 @@
 #  define NOINLINE_DECL(f) f __attribute__((noinline))
 #endif
 
+#if defined(_OS_WINDOWS_) && !defined(_COMPILER_MINGW_)
```
Comment (tkelman) [inline src/support/dtypes.h]: why is this duplicated?

**Comment by iamed2** [comment]
I'm getting compile errors with threading turned on:
```
    CC src/codegen.o
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5350:54: error: no member named 'GeneralDynamicTLSModel' in 'llvm::GlobalValue'
                           NULL, "jl_pgcstack", NULL JL_THREAD_MODEL);
                                                     ^~~~~~~~~~~~~~~
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5343:39: note: expanded from macro 'JL_THREAD_MODEL'
#define JL_THREAD_MODEL ,GlobalValue::GeneralDynamicTLSModel
                         ~~~~~~~~~~~~~^
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5356:66: error: no member named 'GeneralDynamicTLSModel' in 'llvm::GlobalValue'
                           NULL, "jl_exception_in_transit", NULL JL_THREAD_MODEL);
                                                                 ^~~~~~~~~~~~~~~
/Users/ericdavies/repos/juliathreading/src/codegen.cpp:5343:39: note: expanded from macro 'JL_THREAD_MODEL'
#define JL_THREAD_MODEL ,GlobalValue::GeneralDynamicTLSModel
                         ~~~~~~~~~~~~~^
```
I'm on the `jn/tls37` branch and such.

**Comment by kpamnany** [comment]
Threading requires a patched LLVM. Create a ``Make.user`` containing:

```
LLVM_VER = svn
LLVM_GIT_URL_LLVM = https://github.com/JuliaLang/llvm.git -b jn/tls37
LLVM_DEBUG = 0
LLVM_ASSERTIONS = 1
```

**Comment by ranjanan** [comment]
@vtjnash Unfortunately, I keep getting this error:
```
LLVM ERROR: Cannot select: 0x1d7afd0: i64 = X86ISD::WrapperRIP 0x1d7aea8 [ORD=10] [ID=72]boot.jl:290
  0x1d7aea8: i64 = TargetGlobalTLSAddress<%jl_value_t*** @jl_pgcstack> 0 [TF=10] [ORD=10] [ID=42]boot.jl:290
In function: julia__new_0
make[1]: *** [/home/ranjan/julia-threading/usr/lib/julia/inference0.ji] Error 1
make: *** [julia-inference] Error 2
```
@kpamnany which platform did you compile on?

**Comment by iamed2** [comment]
@kpamnany Yes I have that and I get the errors I posted above.

**Comment by kpamnany** [comment]
@iamed2: See https://github.com/JuliaLang/llvm/blob/jn/tls37/include/llvm/IR/GlobalValue.h#L118. Not sure why your build isn't seeing this definition. Do you have an old ``deps/llvm-svn`` hanging around?

@ranjanan, @vtjnash: confirming that LLVM error when trying to build on Linux. Builds on OS X though.

**Comment by vtjnash** [comment]
Linux probably needs us to explicitly request PIC_ (https://github.com/JuliaLang/julia/pull/13410/files#diff-6d4d21428a67320600faf5a1a9f3a16aR6005). It's already the default on the other platforms, since it is invalid to build a shared library without it (as linux is probably trying to do here). I had issues with non-x86_64 builds with that configuration though, so this may be something in investigate further after merging the PR.

**Comment by kpamnany** [comment]
Yeah, changing ``Reloc::Default`` to ``Reloc::PIC_`` at ``codegen.cpp:6005`` fixes this. I suggest making this the default as Linux x86_64 is probably the key platform for all this.

**Comment by vtjnash** [comment]
that would potentially break the normal build on non-X86_64 linux systems, so it'll have to be done with some care

**Comment by JeffBezanson** [comment]
I propose using an `#ifdef` :P

**Comment by kpamnany** [comment]
We need to make it easier for people to try this out. Right now, to enable threading you have to uncomment [JULIA_ENABLE_THREADING](https://github.com/JuliaLang/julia/blob/jn/threading/src/julia.h#L70), and create a ``Make.user`` with ``LLVM_VER`` and ``LLVM_GIT_URL_LLVM`` set correctly, as described above. And for Linux x86_64, edit ``codegen.cpp``, also as described above.

Can we instead ask for a ``Make.user`` with something like ``JULIA_THREADING = 1`` in it? Can the remaining requirements be handled automatically? I'm not sure how ``LLVM_VER`` etc. would be updated -- ``Make.user`` is parsed only after ``Make.inc``, correct?

**Comment by JeffBezanson** [comment]
+1 to that. Make.user is loaded early within Make.inc, so this should be possible.

**Comment by iamed2** [comment]
@kpamnany Nope, I do have that line in my GlobalValue.h. Fresh clone of julia, all the required make options (+ OPENBLAS_TARGET_ARCH=HASWELL). I verified that the correct version of llvm-svn is cloned.

**Comment by kpamnany** [comment]
@iamed2: if you have the correct ``llvm-svn`` then sure, you must have the right ``GlobalValue.h`` in there but the error you're seeing suggests that ``codegen.cpp`` is being compiled with the wrong ``GlobalValue.h``; where is that one? Perhaps in ``$(JULIA_HOME)/usr/include/llvm/IR``? You could try ``make -n`` and look for line in which ``codegen.cpp`` is being compiled to find out.

**Comment by ViralBShah** [comment]
Can we have a Make.user in this branch so that people can try the threading branch easily? Would be great to get it in master with `ifdef` and some basic travis testing.

**Comment by JeffBezanson** [comment]
Let's go with @kpamnany 's idea, and implement a single Make.user flag to enable all of this, then we can just merge this branch into master.

**Comment by iamed2** [comment]
@kpamnany Ah yeah looks like it's pulling in the headers from my Homebrew llvm33-julia. `llvm-config --includedir` returns the correct directory though, so I don't know what's up with that. Seems like a problem in the build process somewhere?

**Comment by tkelman** [comment]
Building the right version of llvm from scratch isn't going to work on travis.

> He says "I don't really have a license, and consider it free to use. If it's going on your website, I'd appreciate a link back to the source (i.e. the link you included in your original email), but that would be a courtesy rather than a requirement." This link is already in lbm3d.m so I think we're good with this.

Ugh. Putting it in our repository is saying we have copyright on the code to distribute it under the MIT license. If we really need to have the matlab code checked in here, we should add in a note indicating we have the author's permission to do this.

**Comment by vtjnash** [comment]
> Ah yeah looks like it's pulling in the headers from my Homebrew llvm33-julia

don't install homebrew in /usr/local: problem solved.

**Comment by iamed2** [comment]
@vtjnash why does the Homebrew install location matter?

**Comment by vtjnash** [comment]
/usr/local is always in the compiler search path, which makes it hard to build applications on that machine that don't have unexpected conflicts or dependencies inherited from homebrew

**Comment by iamed2** [comment]
llvm-config should handle this. The llvm-config built by Julia's build returns the correct locations.

I guess I should make a separate issue for this.

**Comment by iamed2** [comment]
Well, I `brew rm llvm33-julia`d and yet the build (from a fresh clone) is still trying to access it! I'm getting this as part of the build process:

```
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
make[1]: /usr/local/opt/llvm33-julia/bin/llvm-config-3.3: Command not found
```

**Comment by kpamnany** [comment]
@tkelman: added a note in ``lbm3d.m`` stating that we have the author's permission to reproduce the code. Would you go ahead and fix whatever's wrong in ``src/support/dtypes.h``?

@vtjnash and @JeffBezanson: if I did it right (tested on OS X), a ``Make.user`` containing ``JULIA_THREADS = 1`` should be all that is needed to enable threads.

Let's merge this.

**Comment by hayd** [comment]
@kpamnany Perhaps it's worth asking explicitly if the author would willing to allow us to relicense this code under MIT. Their terms appear to be the same as MIT and it would be easier for julia (one less exception).

**Comment by kpamnany** [comment]
@hayd, @tkelman: Author's response is "Hi Kiran. MIT licence agreed. I'm sure this email is enough, but I'll update the page also when I get round to it. Cheers, Iain."

File: src/support/dtypes.h (hunk 7 above)
Comment (tkelman) [inline src/support/dtypes.h]: any reason here, or was this just a wrong merge resolution?

Comment (vtjnash) [inline src/support/dtypes.h]: looks like a merge tool error

**Comment by ViralBShah** [comment]
Bump. Merge this soon?

**Comment by tkelman** [comment]
There's a conflict that needs resolving. I dislike merge commits since they can bury bad conflict resolutions along with thousands of lines of unreviewable diff of everything else that has happened on master, but this is too messy to rebase at this point. A comment about where conflicts were, or a manual commit that only touches the files where conflicts happened would help in tracking conflict resolutions.

**Comment by vtjnash** [comment]
I'm not sure why github doesn't present the merge commits better. `git show` gives me only the lines that were affected by the merge conflict. in the future, I may try to use `git rebase --preserve-merges` to reduce the number of merge commits.

**Comment by kpamnany** [comment]
Yay!

**Comment by ihnorton** [comment]
:joy_cat:

**Comment by johnmyleswhite** [comment]
:tada:

**Comment by tkelman** [comment]
@vtjnash that is a great tip, thanks for that - https://gist.github.com/7e083009a110666331ba

**Comment by vtjnash** [comment]
and even that is a massive over-approximation of the merge conflict (which was just resolving the order of the the assert vs JL_LOCK changes at the top of jl_trampoline_compile_linfo)

**Comment by KristofferC** [comment]
:100:

**Comment by JeffBezanson** [comment]
:fireworks:

**Comment by ViralBShah** [comment]
:+1:

**Comment by tkelman** [comment]
Here's a source build with all the deps in place for this: https://hub.docker.com/r/tkelman/julia-threads/

```
sudo docker run -t -i tkelman/julia-threads /bin/bash
cd /home/julia-x86_64
make -j4 # just julia src and sysimg, all deps are already built
```

**Comment by KristofferC** [comment]
@tkelman Hmm, I seem to get stuck on second round of `inference.jl`..

**Comment by tkelman** [comment]
odd, are you in a VM? maybe adding `MARCH=x86-64` might help?

**Comment by KristofferC** [comment]
Yes, I used the "Kitematic"-thing for Docker on windows which seems to use VirtualBox. Seems to work now with `MARCH`

**Comment by tkelman** [comment]
Probably a virtualbox bug with avx.

**Comment by StefanKarpinski** [comment]
Getting this merge is so awesome.

**Comment by KristofferC** [comment]
Since I am eager to play with this, is there any documentation yet for the new thread-macros or is it best to just look at the examples and go from there?

**Comment by ViralBShah** [comment]
Best to look at the threads perf tests. The interface is very rudimentary. I don't think we should document until APIs stabilise.

**Comment by kpamnany** [comment]
There are a few small example programs in ``test/perf/threads/``. The simplest use cases are in ``test/threads.jl`` which is kind of a tutorial.

The interface is, in fact, rudimentary and the syntax preliminary. However, I think the basic interface (``jl_threading_run(threadfun, args)``) is extremely powerful. Not only does it allow high performance parallel for loops (which might be the single most widely used parallelism construct in technical computing applications designed for scale), but together with atomics, which are also here, it can easily be used to build other shared memory parallelism models, from transactional memory to work-stealing. So, while there are certainly many many things remaining to do, I'm hoping many people will use this and offer feedback to drive priorities, or even better, write some code and raise some PRs.

@ViralBShah, @StefanKarpinski: what's a good venue to discuss the threading interface and improvements to it? Here? In https://github.com/JuliaLang/julia/issues/1790?

**Comment by ViralBShah** [comment]
I think we should close #1790 and open specific issues to discuss various threading related issues and improvements.

**Comment by KristofferC** [comment]
Using this branch I could speed up my kd tree creation with a factor of about 2.25x encountering very little problem. Really nice job here! I look forward to when this is enabled by default.

**Comment by kpamnany** [comment]
@KristofferC: that kd-tree code would be a nice performance benchmark for ``test/perf/threads/`` if you can share it. PRs for performance tests would be very welcome.

**Pull request merged**